### ISSUE TITLE
AppImage builds for Linux

### ISSUE THREAD

**Issue by rottencandy** [STJr/SRB2#457]
Currently the only recommended way to install the game on Linux seems through the Ubuntu ppa, which means that all the other distros have to compile from source. Moreover, the [compilation instructions for linux](https://wiki.srb2.org/wiki/Source_code_compiling/Makefiles) seem to be incomplete(do not mention how to get game assets), making it very difficult for Linux users to get the game.

Providing an [AppImage](https://appimage.org/) would be very beneficial for Linux users since they do not require any additional dependencies, single-file download, and no compilation, among many other advantages.

This can also be directly integrated into Travis, so that new builds are created for every release. 

The process of building an AppImage is also pretty straightforward. Here are the commands I used to build one(tested and works on Debian 10 x64):
```sh
# build the application
mkdir build && cd build
cmake .. -DCMAKE_INSTALL_PREFIX=/usr/bin
make install DESTDIR=AppDir

# Create desktop file
cat > app.desktop <<EOF
[Desktop Entry]
Encoding=UTF-8
Type=Application
Name=SRB2
Comment=Open Source 3D Game
Icon=icon
Exec=AppRun %F
Categories=Game;
EOF
# Get app icon
cp ../src/sdl/SRB2Pandora/icon.png .
# create app entrypoint
cat > AppDir/AppRun <<EOF
#!/usr/bin/sh
HERE="$(dirname "$(readlink -f "${0}")")"
exec $HERE/usr/bin/lsdlsrb2-2.2.2
EOF
chmod +x AppDir/AppRun

# Build AppImage
curl -L https://github.com/linuxdeploy/linuxdeploy/releases/download/continuous/linuxdeploy-x86_64.AppImage -o linuxdeploy
chmod +x linuxdeploy
./linuxdeploy --appdir AppDir --output appimage -d app.desktop -i icon.png
```
I could create a PR for this myself, but couldn't figure out which scripts are run by the Travis deployer bots.

**Comment by 4ifi** [comment]
@rottencandy, I improved your build script: (https://github.com/4ifi/appimages/blob/master/srb2_build_appimage.sh)

**Comment by rottencandy** [comment]
Thanks @4ifi , It would be great if your work was merged into SRB's CI/CD pipelines. Would benefit lots of Linux users. I'll try to look into the travis files and see if I can figure out how we can integrate your scripts.

**Comment by mazmazz** [comment]
This in my radar. Thanks for your hard work @rottencandy @4ifi . 

I might be re-working the Travis bots soon, and I want to get rid of the current Debian scripts. Does this work depend on those scripts? 

Instead of working off the current bots, it may be more productive if you got AppImage builds working on Travis from scratch. If you showed me that, it would be a big help so I know what to incorporate.

**Comment by 4ifi** [comment]
@mazmazz, it doesn't depend on debian. You can fork and easily rewrite script for arch linux or smth else. No matter how you download build essentials and libjack. It only depends on bash syntax. So you can run prepared appimage from every linux distro except android & chrome os & other crap. But I know travis ci currently supports ubuntu.

**Comment by mazmazz** [comment]
@4ifi what's the reason for bundling `libjack-jackd2-0` in your script? I'm not primarily a Linux guy, so I'm unsure if we actually require this library. @SteelTitanium ?

**Comment by SteelTitanium** [comment]
I'm not entirely sure if that's required, SDL2 do have multiple audio drivers, and in most cases usually picks the best one for the system.

**Comment by 4ifi** [comment]
@mazmazz @SteelTitanium , thanks for your reply. I couldn't launch SRB2 on debian server minimal installation with Xorg because of libjack-jackd2-0 was missing!!! So I included this into AppImage. SRB2 AppImage must run on every system.

**Comment by SteelTitanium** [comment]
One thing I wonder, does it only support PNG files for the icon, does it support XPM files, Because if so. I rather that be used

**Comment by SteelTitanium** [comment]
The icon file used is outdated, and if there's plans on making the appimage official, I rather the current icon be used as  representative of it

**Comment by 4ifi** [comment]
@SteelTitanium replaced icon with blank file in my script.

**Comment by 4ifi** [comment]
@SteelTitanium You're allowed to view code of AppDir to AppImage software because I don't understand C++ code: https://github.com/linuxdeploy/linuxdeploy
If XPM isn't supported you can convert XPM to PNG.

**Comment by 4ifi** [comment]
@SteelTitanium In my opinion XPM in linux deploy should be supported. Why not. Just fork and test.

**Comment by rottencandy** [comment]
The XDG icon spec actually recommends SVG icons because they scale really well for different icon sizes, but XPM is also supported, and the AppDir can optionally also contain a set of differently sized icons(e.g., 64x64, 128x128, etc.)
https://docs.appimage.org/reference/appdir.html#general-description

**Comment by mazmazz** [comment]
I need help determining which libraries we need to include in the AppImage. These are our current depends:

```
 libsdl2-2.0-0,
 libsdl2-mixer-2.0-0,
 zlib1g,
 libgme0,
 libopenmpt | libopenmpt0,
 libpng | libpng16-16 | libpng12-0
```

Those depends also have depends, obviously. I've heard two things about how far we need to go: either we assume that certain libraries are installed by default, or [we include everything down to `ld-linux.so`](https://discourse.appimage.org/t/does-appimage-include-all-shared-library-dependencies-by-default/509/6).

If any of you can come up with an inclusive list of libraries, I would be grateful.

Whichever libraries we include, I believe we need to compile with `CFLAGS=--march=nocona --mtune=haswell` like I intend to do with future game releases.

**Comment by mazmazz** [comment]
If you'd like to try a pull request, [you may work off the `dpl-2` branch](https://github.com/STJr/SRB2/compare/dpl-2):

* [.travis.yml](https://github.com/STJr/SRB2/blob/dpl-2/.travis.yml)
* [deployer.sh:217](https://github.com/STJr/SRB2/blob/dpl-2/travis/deployer.sh#L217) -- This is where OS X and Windows installers are generated.

**Comment by SteelTitanium** [comment]
On the post after that post, there's a link to libraries that are excluded by default, and ld-linux.so is one of them

**Comment by mazmazz** [comment]
I realize that [`linuxdeploy`](https://github.com/linuxdeploy/linuxdeploy) already takes care of this problem. It pulls dependencies and bundles them into the AppImage.

That also means our compatibility is limited to however those depends' binaries were built. I suppose, for a simple solution, we can choose not to worry about building libraries.

**Comment by rottencandy** [comment]
@mazmazz tested your build, I guess the tcl wrapper lib(`libwrap.so.0`) was missing, I installed it and was able to launch run the AppImage without any issues.

Currently running Fedora 32 btw.

**Comment by 4ifi** [comment]
Devuan server core iso installation with Xorg & video drivers and without DE:
/tmp/.mount_srb2-lSXg5yU/usr/bin/lsdlsrb2-2.2.4: error while loading shared libraries: libjack.so.0: cannot open shared object file: No such file or directory

**Comment by 4ifi** [comment]
@mazmazz AppImage must include all libs which SRB2 needs

**Comment by rottencandy** [comment]
> /tmp/.mount_srb2-lSXg5yU/usr/bin/lsdlsrb2-2.2.4: error while loading shared libraries: libjack.so.0: cannot open shared object file: No such file or directory
> /tmp/.mount_srb2-llrgCu2/usr/bin/lsdlsrb2-2.2.4: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory

@4ifi could it be that `libjack` is a Deuan-only dependency? I don't have any such lib on my both my Fedora and Debian systems, but I was able to run the AppImage without any errors.

**Comment by 4ifi** [comment]
@rottencandy libjack-dev doesn't include libjack.so.0 library, maybe you meant libjack-jackd2-0 package. Devuan is a Debian fork with banned *systemd* repository packages. Debian full installatioln includes a lot of libraries by default. I don't think if you installed Debian in partiotion with debootstrap & grub-install like pro linux guy. Don't touch libjack.so.0 library.

**Comment by mazmazz** [comment]
@4ifi I didn't include `libjack` because I think I tried including that lib and it caused crashes on Ubuntu systems. I don't remember. It's also reasonable to assume that the user has a working audio system before installing SRB2.

The adage that AppImage bundles every required library is a marketing gimmick. In practice, it doesn't really work because certain libraries can cause crash conflicts in certain systems, like my previous replies show. 

I really don't know much about Linux installs; sorry I can't help further. I asked the Discord if someone will finish up this work, because I stepped off the project.

**Comment by rottencandy** [comment]
@mazmazz I think your present work is really great and works fine without any issues, I would say that it is in a playable state. The AppImage should not try and include every single possible dependent library as you said.

I believe it would be much more beneficial if the current builds are provided to more Linux users so they could try it out and provide feedback, which would help decide whether or not to add more dependencies to the build.

What do you think @4ifi ?

**Comment by mazmazz** [comment]
I've updated the AppImage to v2.2.5, and also added `libwrap` and `libtinfo` to the bundle.

https://github.com/mazmazz/SRB2/releases/download/deployer-20200710b/srb2-linux-deployer-2-appimage-706946514-2c2e91f0.tar.gz

**Comment by mazmazz** [comment]
We got a `libpng12` missing error from Ubuntu 20.04. 

I don't remember if I took any steps to fix this, but I'm not in a position to resolve it right now.

**Comment by mazmazz** [comment]
Error when trying to download WADs from a netgame server

This happens because the game tries to write to the AppImage, which is read-only.

I'm not sure if there's a consvar or specific behavior we can trigger to point the game to a writable directory?

![image](https://user-images.githubusercontent.com/297483/87211548-0298f000-c2e8-11ea-95f1-855e5cfe38a1.png)

**Comment by rottencandy** [comment]
> I'm not sure if there's a config variable or specific behavior we can trigger to point the game to a writable directory?
The game seems to be able to read and write to `~/.srb2`, isn't that where it stores WADs as well?

Wasn't able to test downloading WADs because the netgame server would not allow anything less than v2.2.6

**Comment by mazmazz** [comment]
I thought so too, but I think the issue is that the program is trying to check free space. It doesn't check the free space of the volume containing `~/.srb2`, but instead it checks that of the AppImage volume.

**Comment by HazelDust** [comment]
Thanks for putting this together, I've been using this appimage as I had difficulty getting audio to work using the flatpak (running Linux Mint 20).

Everything seems to works fine except I can't figure out how to mod in 3D models (specifically Jeck Jims' 2.2 3D Models v1.1). I tried rebuilding the appimage with the model.dat and model directory in the bin directory but the 3D models won't show up in game. When I run via terminal I see the error message "Error while loading models.dat: No such file or directory". Am I putting them in the right place?

**Comment by Guiazsan** [comment]
> Thanks for putting this together, I've been using this appimage as I had difficulty getting audio to work using the flatpak (running Linux Mint 20).
> 
> Everything seems to works fine except I can't figure out how to mod in 3D models (specifically Jeck Jims' 2.2 3D Models v1.1). I tried rebuilding the appimage with the model.dat and model directory in the bin directory but the 3D models won't show up in game. When I run via terminal I see the error message "Error while loading models.dat: No such file or directory". Am I putting them in the right place?

i finally find where do you need to put, its inside the .srb2 in your home folder

**Comment by GoldenTails** [comment]
Hello! Sorry for getting back here so late. I've implemented a method to create working SRB2 AppImage builds for the Linux community.
I now maintain (and have been maintaining) AppImage releases here: https://github.com/GoldenTails/SRB2/releases

For those interested, linked below is the code I use to create AppImages for SRB2, based upon both SRB2's `master` and `next` branches:
https://github.com/GoldenTails/SRB2/tree/appimage-master
https://github.com/GoldenTails/SRB2/tree/appimage-next

The AppImage scripts require Python to be installed as I type here, so go grab that.

To create AppImages through Makefiles, include `APPIMAGE=1` in your `make` command. Sorry CMake people -- there is currently no way to do this directly via CMake, however you can use the included `AppImage_build.sh` with the correct parameters to manually build an AppImage.

If you don't include `APPIMAGE_NOASSETS=1` into your `make` command, attempts will be made to locate SRB2's assets within `assets/installer` relative to the repository root. Those manually using `AppImage_build.sh` should include the `--small` parameter to evoke the same behavior as `APPIMAGE_NOASSETS=1`.

If `appimagetool` is not found within your `PATH`, an x86_64 copy will be downloaded temporarily to assist with creating AppImages, and will be removed afterwards as part of cleanup. I highly recommend installing `appimagetool` for your architecture if you want to save on bandwidth and alleviate issues with architecture incompatibilities.

Leaving this issue open as while this is definitely allows the creation of AppImage builds for Linux, I am independently and unofficially maintaining this, and opening an issue on the main SRB2 repo implies the need for some sort of official support before this can be truly resolved.

Also my implementation isn't perfect -- the way the Makefile handles AppImage builds is suboptimally implemented and the scripts themselves need some refinement too. ex: `APPIMAGE_NOASSETS=1` in the Makefile will automatically define `APPIMAGE`, python may not be required, architecture shouldn't be assumed when downloading `appimagetool`, etc.

tl;dr Linked above is my code to make AppImages, but it wouldn't make sense to close this issue until it becomes officially supported and hopefully cleaned up.
